File: 이건희/SWexpertacademy/D3/4047.py
T = int(input())
for test_case in range(1, T + 1):
    T = input()
    S = {}
    H = {}
    D = {}
    C = {}
    result = 0
    for i in range(0, len(T), 3):
        pattern = T[i]
        num = T[i + 1 : i + 3]
 
        if pattern == 'S':
            if num in S:
                result = 1
                break
            S[num] = 1
 
        elif pattern == 'H':
            if num in H:
                result = 1
                break
            H[num] = 1
 
        elif pattern == 'D':
            if num in D:
                result = 1
                break
            D[num] = 1
 
        elif pattern == 'C':
            if num in C:
                result = 1
                break
            C[num] = 1
         
    print(f'#{test_case} {13 - len(S)} {13 - len(D)} {13 - len(H)} {13 - len(C)}') if result == 0 else print(f'#{test_case} ERROR')

File: 이훈일/삼성 SW 역량 테스트 기출 문제/D3/4047.java
import java.util.Scanner;
import java.io.FileInputStream;
import java.util.*;

class Solution {
	public static void main(String args[]) throws Exception {
		Scanner sc = new Scanner(System.in);
		int T;
		T=sc.nextInt();
	
		for(int test_case = 1; test_case <= T; test_case++) {
			String line = sc.next();
            HashMap<String, boolean[]> cards = new HashMap<String, boolean[]>(){{
            	put("S", new boolean[13]);
                put("C", new boolean[13]);
                put("H", new boolean[13]);
                put("D", new boolean[13]);
            }};
           
            boolean isPossible = true;
            for(int i = 0; i < line.length(); i += 3) {
            	String sub = line.substring(i, i+3);
                String type = String.valueOf(sub.charAt(0));
                int num = Integer.valueOf(sub.substring(1, 3));
                if(cards.get(type)[num-1]) {
                	isPossible = false;
                    break;
                } else {
                	boolean[] cardsCalled = cards.get(type);
                    cardsCalled[num-1] = true;
                    cards.put(type, cardsCalled);
                }
            }
            
            String answer = "";
            String[] types = new String[]{"S", "D", "H", "C"};
            if(isPossible) {
            	for(String key : types) {
                    int count = 0;
                	for(boolean isCalled : cards.get(key)) {
                    	if(!isCalled) {
                        	count++;
                        }
                    }
                    answer += count+" ";
                }
            } else {
            	answer = "ERROR";
            }
            
            System.out.println("#" + test_case + " " + answer);
		}
	}
}
